domain/modelconfig/service/service.go

```diff
@@ -26,6 +28,13 @@ type ModelDefaultsProvider interface {
 	ModelDefaults(context.Context) (modeldefaults.Defaults, error)
 }
 
+// ModelConfigProviderFunc describes a type that is able to return a
+// [environs.ModelConfigProvider] for the specified cloud type. If the no
+// model config provider exists for the supplied cloud type then a
+// [coreerrors.NotFound] error is returned. If the cloud type provider does not
+// support model config then a [coreerrors.NotSupported] error is returned.
+type ModelConfigProviderFunc func(string) (environs.ModelConfigProvider, error)
```

Comment:
It's a shame that `func(string)` isn't typed, otherwise we wouldn't have gotten ModelType and CloudType mixed up.

Reply:
I would propose a name change here only because provider is very overload and is a big scope. Maybe something like `ModelProviderConfigSchemaFunc`.